src/hb-buffer.h

```diff
@@ -76,18 +76,63 @@ typedef struct hb_glyph_info_t {
  * @HB_GLYPH_FLAG_UNSAFE_TO_BREAK: Indicates that if input text is broken at the
  * 				   beginning of the cluster this glyph is part of,
  * 				   then both sides need to be re-shaped, as the
- * 				   result might be different.  On the flip side,
- * 				   it means that when this flag is not present,
- * 				   then it's safe to break the glyph-run at the
- * 				   beginning of this cluster, and the two sides
- * 				   represent the exact same result one would get
- * 				   if breaking input text at the beginning of
- * 				   this cluster and shaping the two sides
- * 				   separately.  This can be used to optimize
- * 				   paragraph layout, by avoiding re-shaping
- * 				   of each line after line-breaking, or limiting
- * 				   the reshaping to a small piece around the
- * 				   breaking point only.
+ * 				   result might be different.
+ *
+ * 				   On the flip side, it means that when this
+ * 				   flag is not present, then it is safe to break
+ * 				   the glyph-run at the beginning of this
+ * 				   cluster, and the two sides will represent the
+ * 				   exact same result one would get if breaking
+ * 				   input text at the beginning of this cluster
+ * 				   and shaping the two sides separately.
+ *
+ * 				   This can be used to optimize paragraph
+ * 				   layout, by avoiding re-shaping of each line
+ * 				   after line-breaking.
+ *
+ * @HB_GLYPH_FLAG_UNSAFE_TO_CONCAT: Indicates that if input text is changed on one
+ * 				   side of the beginning of the cluster this glyph
+ * 				   is part of, then the shaping results for the
+ * 				   other side might change.
+ *
+ * 				   Note that the absence of this flag will NOT by
+ * 				   itself mean that it IS safe to concat text.
+ * 				   Only two pieces of text both of which clear of
+ * 				   this flag can be concatenated safely.
+ *
+ * 				   This can be used to optimize paragraph
+ * 				   layout, by avoiding re-shaping of each line
+ * 				   after line-breaking, by limiting the
+ * 				   reshaping to a small piece around the
+ * 				   breaking point only, even if the breaking
+ * 				   point carries the
+ * 				   #HB_GLYPH_FLAG_UNSAFE_TO_BREAK or when
+ * 				   hyphenation or other text transformation
+ * 				   happens at line-break point, in the following
+ * 				   way:
+ *
+ * 				   1. Iterate back from the line-break point till
+ * 				   the the first cluster start point that is
+ * 				   NOT unsafe-to-concat, 2. shape the segment from
+ * 				   there till the end of line, 3. check whether the
+ * 				   resulting glyph-run also is clear of the
+ * 				   unsafe-to-concat at its start-of-text position;
+ * 				   if it is, just splice it into place and the line
+ * 				   is shaped; If not, back up to a point further back
+ * 				   that is clear of unsafe-to-concat and retry.
+ *
+ * 				   At the start of next line a similar algorithm can
```

Comment:
I am trying to grasp this paragraph. Does this mean for computing a good line break that keeps line-breaking free of any shaping changes, one would have to find a candidate both by checking unsafe-to-concat "join compatibility" with the current line as described above, then check the segment after the candidate position for compatibility with the next line - and that's where the problem arises of being unable to check unsafe-to-concat state on the end of the run?

Reply:
We first shape the full paragraph. Then our aim is to reuse the shaping result of the paragraph to produce shaping result for each line. But the line is being broken and we need to patch the line start / end for proper shaping results, right? The algorithm I documented explains what to do for the end of line. But when you start the next line, you need to reshape the beginning of the line. That's where you need to find a safe-to-concat point and splice in a piece. Does that make sense?

Reply:
> That's where you need to find a safe-to-concat point and splice in a piece. Does that make sense?

Only if the line-break position wasn't safe-to-break.

Reply:
I clarified the documentation a bit. Please check.

Reply:
> At the start of next line a similar algorithm can be implemented.

I still don't get this line. What does "similar" mean here if we spelled it out. What do we want to achieve for the beginning of the line in terms of safe-to-concat "clean"-ness.

Reply:
Just added this:
https://github.com/harfbuzz/harfbuzz/pull/3297/commits/540af548dc3926c4b9db059c9b20297df0802671

Reply:
Thanks, yes, I find that more clear.